### ISSUE TITLE
[feature request] type guard return types should throw error when type of argument is wrong type

### ISSUE THREAD

**Issue by trusktr** [Microsoft/TypeScript#29162]
## Suggestion

The following code makes no sense because the typeof `b` inside the conditional block is `string & number`:

```ts
function isNumber(num: any): num is number {
    return typeof num === 'number'
}

const b: string = 'asdf'

if (isNumber(b)) {
    b.toExponential(2) // typeof b is `string & number`, what???
}
```

[playground link](http://www.typescriptlang.org/play/#src=function%20isNumber(num%3A%20any)%3A%20num%20is%20number%20%7B%0A%20%20%20%20return%20typeof%20num%20%3D%3D%3D%20'number'%0A%7D%0A%0Aconst%20b%3A%20string%20%3D%20'asdf'%0A%0Aif%20(isNumber(b))%20%7B%0A%20%20%20%20b.toExponential(2)%20%2F%2F%20typeof%20b%20is%20%60string%20%26%20number%60%2C%20what%3F%3F%3F%0A%7D)

## Use Cases

It should just throw a type error, because if it clearly knows we have a string, then due to the type signature of `isNumber`, it should obviously fail.

Otherwise the `is` keyword doesn't seem to be doing its job fully.

## Checklist

My suggestion does not meet all these guidelines:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code. **It may be a small breaking change, as some code that doesn't throw before may now throw.**
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).
